# CNVPipe - an integrated and robust CNV calling pipeline

CNVPipe integrates several popular CNV calling methods and designed various score metrics to filter false positive callings, aims at reaching higher sensitivity and lower false discovery rate. CNVPipe uses Snakemake as bachend to it highly robust and reproducible between different machines. CNVPipe is user-friendly and robust.

## CNVPipe workflow

CNVPipe accept fastq file format as input, then perform the reads quality control by [fastp](https://github.com/OpenGene/fastp) and reads alignment with [bwa mem](http://bio-bwa.sourceforge.net/bwa.shtml). It will automatically determine an optimal CNV calling resolution if users not provide one. Then CNVPipe parallelly call CNVs by using five methods, including [CNVKit](https://github.com/etal/cnvkit), [CNVpytor](https://github.com/abyzovlab/CNVpytor), [cn.MOPS](http://bioconductor.org/packages/devel/bioc/html/cn.mops.html), [Smoove](https://github.com/brentp/smoove) and [Delly](https://github.com/dellytools/delly). After that, CNVPipe will adopt different strategies to recursively merge the CNV set from different methods based on the read depth. To obtain high-confidence CNVs, CNVPipe firstly use [duphold](https://github.com/brentp/duphold) to calculate the adjacent read depth of CNVs and assign "duphold score" for each CNV entry. Secondly, CNVPipe use [CNVFilteR](http://bioconductor.org/packages/release/bioc/html/CNVfilteR.html) to filter likely false positive CNVs based on B allele frequency of SNPs. In addition, CNVPipe calculate the overlap length between identified CNVs and genomic low-complexity regions to further refine the CNV set. To predict the pathogenicity of CNVs, CNVPipe compare the identified CNVs with CNVs frequently (>1%) appeared in healthy population, and assign a score based on the overlap fraction. CNVPipe also utilize [ClassifyCNV](https://github.com/Genotek/ClassifyCNV) to do a pathogenicity prediction. Finally CNVPipe will give a CNV list with various score metrics, predicted pathogenicity and possibly affected dosage-sensitive genes, and also give a list of figures showing read depth and BAF of high-confident CNVs.

The following figure demonstrates the workflow of CNVPipe.

![CNVPipe workflow](/doc/logo/CNVPipe-workflow.png)